### ISSUE TITLE
Check panics early when run in a loop

### ISSUE THREAD

**Issue by sargunv** [de-vri-es/assert2-rs#23]
Hi, I found this library when looking for a way to run tests with multiple assertions that all evaluate and report errors before panicking at the end. From reading your docs, it appears that the `check!` macro in this library is intended to do that. However, it doesn't appear to be working for me when run in a loop (I assume it only works within a single scope?). Here's my use case:

I'm writing a program designed to process a builtin config with many different "command" definitions, and for ease of use the tests for each command are embedded in the same config. I'd like my test to run and report on each "test" in the config, so one failing test doesn't stop the rest from being run.

![image](https://user-images.githubusercontent.com/1320357/136739660-9e47d56b-dd4a-4fa6-882c-a8c8d47ccb53.png)

I've deliberately tweaked the above config the ensure two different tests fail. However, when running the tests, only one assertion failed is reported before the test panics. See my test code and test output below. You can see the `check!` is run in a loop.

![image](https://user-images.githubusercontent.com/1320357/136739780-295b67f0-c72a-4299-8353-9070513755b0.png)

I tried putting two `check!` calls back to back in the same scope, and that did work as expected; both reported failure before the panic. So, it appears multiple `check!` calls are limited to running together only if they exist in the same scope.

I assume this is by design and not a bug, but I'd love a feature in this repo that lets me check and report on multiple conditions dynamically within a loop regardless of scope.

**Comment by de-vri-es** [comment]
You are correct that this only works within the same scope. However, I also totally agree with the behaviour you want. I want it too ;)

The biggest problem now is that we *must* panic to fail a test. The best thing we can do is delay the panic until a later point. So the `check!()` macro does that with a scope guard that panics when dropped. I can't think of anything better currently :(

Something like `::test::fail()` to fail the current test without panicking would help, that that requires support from the standard library. I'm open to other suggestions too, but keep in mind it should work correctly with asynchronous code too.

**Comment by DanielJoyce** [comment]
You could have a version of check that returned the scope guard so then you could decide when to drop it?
